pkg/virt-handler/heartbeat/ksm.go

```diff
@@ -0,0 +1,299 @@
+ *
+ */
+
+package heartbeat
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/log"
+
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+)
+
+const (
+	pagesBoostDefault      = 300
+	pagesDecayDefault      = -50
+	nPagesMinDefault       = 64
+	nPagesMaxDefault       = 1250
+	nPagesInitDefault      = 100
+	sleepMsBaselineDefault = 100 // 10ms in oVirt seemed really low
+	freePercentDefault     = 0.2
+)
+
+var (
+	// These are vars so they can be changed by the unit tests
+
+	// In some environments, sysfs is mounted read-only even for privileged
+	// containers: https://github.com/containerd/containerd/issues/8445.
+	// Use the path from the host filesystem.
+	ksmBasePath  = "/proc/1/root/sys/kernel/mm/ksm/"
+	ksmRunPath   = ksmBasePath + "run"
+	ksmSleepPath = ksmBasePath + "sleep_millisecs"
+	ksmPagesPath = ksmBasePath + "pages_to_scan"
+
+	memInfoPath = "/proc/meminfo"
+
+	pagesBoost              = pagesBoostDefault
+	pagesDecay              = pagesDecayDefault
+	nPagesMin               = nPagesMinDefault
+	nPagesMax               = nPagesMaxDefault
+	nPagesInit              = nPagesInitDefault
+	sleepMsBaseline uint64  = sleepMsBaselineDefault // 10ms in oVirt seemed really low
+	freePercent     float32 = freePercentDefault
+)
+
+type ksmState struct {
+	running bool
+	sleep   uint64
+	pages   int
+}
+
+// Inspired from https://github.com/artyom/meminfo
+func getTotalAndAvailableMem() (uint64, uint64, error) {
+	var total, available uint64
+
+	f, err := os.Open(memInfoPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer f.Close()
+	s := bufio.NewScanner(f)
+	found := 0
+	for s.Scan() && found < 2 {
+		switch {
+		case bytes.HasPrefix(s.Bytes(), []byte(`MemTotal:`)):
+			_, err = fmt.Sscanf(s.Text(), "MemTotal:%d", &total)
+			found++
+		case bytes.HasPrefix(s.Bytes(), []byte(`MemAvailable:`)):
+			_, err = fmt.Sscanf(s.Text(), "MemAvailable:%d", &available)
+			found++
+		default:
+			continue
+		}
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+	if found != 2 {
+		return 0, 0, fmt.Errorf("failed to find total and available memory")
+	}
+
+	return total, available, nil
+}
+
+func getKsmPages() (int, error) {
+	pagesBytes, err := os.ReadFile(ksmPagesPath)
+	if err != nil {
+		return 0, err
+	}
+
+	pages, err := strconv.Atoi(strings.TrimSpace(string(pagesBytes)))
+	if err != nil {
+		return 0, err
+	}
+
+	return pages, nil
+}
+
+// Inspired from https://github.com/oVirt/mom/blob/master/doc/ksm.rules
+func calculateNewRunSleepAndPages(running bool) (ksmState, error) {
+	ksm := ksmState{running: running}
+	total, available, err := getTotalAndAvailableMem()
+	if err != nil {
+		return ksm, err
+	}
+	ksm.pages, err = getKsmPages()
+	if err != nil {
+		return ksm, err
+	}
+
+	ksm.sleep = sleepMsBaseline * (16 * 1024 * 1024) / (total - available)
```

Comment:
Can you explain the `ksm.sleep` calculations? A comment would be great.

Reply:
Comment added